### PR TITLE
doc: how to clean up orphaned encrypted_value records

### PR DESCRIPTION
**Pull request by hsinn0**
[#182121168]

Documenting how to cleanup orphaned records as noted in https://github.com/cloudfoundry/credhub/issues/231. Part of SQL scripts here were borrowed from the issue.
Added a section about the orphaned data cleanup steps in `README.md`.

### CODE REVIEW

**Comment by peterhaochen47** [comment]
Could we maybe add this instruction in the ["/docs" directory](https://github.com/cloudfoundry/credhub/tree/830bb16eade21e98cc74b6916419f98fa279a719/docs), and maybe put a link to it in the main README?

I feel that putting a lengthy instruction on a specific issue in the main README might crowd out the high-level overview info that people are typically looking for in a main repo README.

**Comment by Tallicia** [comment]
I agree this should be links from the readme, 1 for each database separately.

**Comment by hsinn0** [comment]
> Could we maybe add this instruction in the ["/docs" directory](https://github.com/cloudfoundry/credhub/tree/830bb16eade21e98cc74b6916419f98fa279a719/docs), and maybe put a link to it in the main README?

Moved to separate doc.

**Comment by hsinn0** [comment]
> I agree this should be links from the readme, 1 for each database separately.

Moved but kept them in a single doc.

File: docs/orphaned-encryption-value-cleanup.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+## Cleaning up orphaned encrypted_value records
+CredHub version 2.12.66 and ealier had a bug where associated `encrypted_value`
+records were not deleted when credentials were deleted.
+(See https://github.com/cloudfoundry/credhub/issues/231.)
+
+To clean up those orphaned encrypted_value records accumulated in the old
+credhub versions, you can run SQL script for your database type as
+described below.
```
Comment (emalm) [inline docs/orphaned-encryption-value-cleanup.md:8]:
```suggestion
CredHub version 2.12.66 and earlier had a bug where `encrypted_value`
records were not deleted when the associated credentials were deleted.
(See https://github.com/cloudfoundry/credhub/issues/231.)

To clean up the orphaned `encrypted_value` records that accumulated
on these older CredHub versions, run the SQL script below for your database type.
```

File: docs/orphaned-encryption-value-cleanup.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+### For postgresql database
```
Comment (emalm) [inline docs/orphaned-encryption-value-cleanup.md:10]:
```suggestion
### For PostgreSQL
```

File: docs/orphaned-encryption-value-cleanup.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+``` sql
+-- Step 1: Create indexes on the 3 related tables
+CREATE INDEX CONCURRENTLY IF NOT EXISTS credential_version_encrypted_value_uuid_idx ON credential_version USING btree (encrypted_value_uuid);
+CREATE INDEX CONCURRENTLY IF NOT EXISTS password_credential_password_parameters_uuid_idx ON password_credential USING btree (password_parameters_uuid);
+CREATE INDEX CONCURRENTLY IF NOT EXISTS user_credential_password_parameters_uuid_idx ON user_credential USING btree (password_parameters_uuid);
+
+-- Step 2: Optoinal. Get the number of encrypted_value records that are not in any of those 3 tables
```
Comment (emalm) [inline docs/orphaned-encryption-value-cleanup.md:17]:
```suggestion
-- Step 2: Optional. Get the number of encrypted_value records that are not in any of those 3 tables
```

File: docs/orphaned-encryption-value-cleanup.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+-- Step 4: Drop the indexes
+DROP INDEX CONCURRENTLY IF EXISTS credential_version_encrypted_value_uuid_idx;
+DROP INDEX CONCURRENTLY IF EXISTS password_credential_password_parameters_uuid_idx;
+DROP INDEX CONCURRENTLY IF EXISTS user_credential_password_parameters_uuid_idx;
+
+```
+Note that adding indexes and running the queries may take very long time if you
+have large database. You may want to run each step one by one before proceeding
+to next step, to get the number of orphaned records before proceeding to
+next step for example.
```
Comment (emalm) [inline docs/orphaned-encryption-value-cleanup.md:44]:
```suggestion
Note that adding indexes and running the queries may take a very long time if you
have a large CredHub database. You may want to run each query separately. In
particular, you may want to count the number of orphaned records first to
determine whether cleaning them up will reduce the size of your CredHub
database significantly.
```

File: docs/orphaned-encryption-value-cleanup.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+### For mysql database
```
Comment (emalm) [inline docs/orphaned-encryption-value-cleanup.md:46]:
```suggestion
### For MySQL or MySQL-compatible databases, such as MariaDB
```

File: docs/orphaned-encryption-value-cleanup.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+``` sql
+-- Step 1: Optoinal. Get the number of encrypted_value records that are not in any of those 3 tables
```
Comment (emalm) [inline docs/orphaned-encryption-value-cleanup.md:48]:
```suggestion
-- Step 1: Optional. Get the number of encrypted_value records that are not in any of those 3 tables
```

File: docs/orphaned-encryption-value-cleanup.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+NOT EXISTS (select 1 from credential_version where encrypted_value_uuid=encrypted_value.uuid)
+AND
+NOT EXISTS (select 1 from password_credential where password_parameters_uuid=encrypted_value.uuid)
+AND
+NOT EXISTS (select 1 from user_credential where password_parameters_uuid=encrypted_value.uuid);
+```
+For MySQL database, the necessary indexes are already there, so you do not have
+to create them. The queries still may take a long time to run if you have large
+database. You may want to get the number of orphaned records before proceeding
+to next step.
```
Comment (emalm) [inline docs/orphaned-encryption-value-cleanup.md:69]:
```suggestion
For MySQL database, the necessary indexes are already there, so you do not have
to create them. The queries still may take a long time to run if you have a large
database. You may want to get the number of orphaned records before proceeding
to the deletion step.
```

File: README.md [hunk 8]
```diff
@@ -219,3 +219,7 @@ CREDENTIAL_ROOT=/path/to/credhub/repo/plus/src/test/resources ./scripts/run_test
 
 Assuming it works, that will generate some test client certificates for testing mutual TLS (in `certs/` in the acceptance test directory) and run the acceptance test suite against your locally running credhub server.
 
+### Cleaning up orphaned encrypted_value records
+To clean up orphaned `encrypted_value` records from CredHub version 2.12.66 and
+ealier (https://github.com/cloudfoundry/credhub/issues/231), follow the steps decribed in
+[Cleaning up orphaned encrypted_value records](docs/orphaned-encryption-value-cleanup.md).
```
Comment (emalm) [inline README.md:225]:
```suggestion
### Cleaning up orphaned encrypted_value records
To clean up orphaned `encrypted_value` records from CredHub version 2.12.66 and
earlier (https://github.com/cloudfoundry/credhub/issues/231), follow the steps decribed in
[Cleaning up orphaned encrypted_value records](docs/orphaned-encryption-value-cleanup.md).
```

**Review by emalm — CHANGES_REQUESTED** [review]
Looks good to me! I just had a few minor adjustments to the surrounding explanatory text.

**Review by emalm — APPROVED** [review]

**Pull request merged**